Sources/Citadel/OpenSSHKey.swift

```diff
@@ -0,0 +1,223 @@
+import CCryptoBoringSSL
+import BigInt
+import Foundation
+import NIO
+import Crypto
+import NIOSSH
+
+internal protocol ReadableFromBuffer {
+    static func read(from buffer: inout ByteBuffer) throws -> Self
+}
+
+internal protocol OpenSSHKeyProtocol {
+    static var keyType: OpenSSH.KeyType { get }
+    associatedtype PublicKey: NIOSSHPublicKeyProtocol
+    associatedtype PrivateKey: NIOSSHPrivateKeyProtocol, ReadableFromBuffer
+}
+
+extension OpenSSHKeyProtocol {
+    static var keyType: OpenSSH.KeyType {
+        .init(rawValue: PrivateKey.keyPrefix)!
+    }
+}
+
+extension Insecure.RSA: OpenSSHKeyProtocol {}
+extension ED25519: OpenSSHKeyProtocol {}
+
+extension Insecure.RSA.PrivateKey: ReadableFromBuffer {
+    static func read(from buffer: inout ByteBuffer) throws -> Self {
+        guard
+            let nBytesLength = buffer.readInteger(as: UInt32.self),
+            let nBytes = buffer.readBytes(length: Int(nBytesLength)),
+            let eBytesLength = buffer.readInteger(as: UInt32.self),
+            let eBytes = buffer.readBytes(length: Int(eBytesLength)),
+            let dLength = buffer.readInteger(as: UInt32.self),
+            let dBytes = buffer.readBytes(length: Int(dLength)),
+            let iqmpLength = buffer.readInteger(as: UInt32.self),
+            let _ = buffer.readData(length: Int(iqmpLength)),
+            let pLength = buffer.readInteger(as: UInt32.self),
+            let _ = buffer.readData(length: Int(pLength)),
+            let qLength = buffer.readInteger(as: UInt32.self),
+            let _ = buffer.readData(length: Int(qLength))
+        else {
+            throw InvalidKey()
+        }
+        
+        let privateExponent = CCryptoBoringSSL_BN_bin2bn(dBytes, dBytes.count, nil)!
+        let publicExponent = CCryptoBoringSSL_BN_bin2bn(eBytes, eBytes.count, nil)!
+        let modulus = CCryptoBoringSSL_BN_bin2bn(nBytes, nBytes.count, nil)!
+
+        return self.init(privateExponent: privateExponent, publicExponent: publicExponent, modulus: modulus)
+    }
+}
+
+extension ED25519.PrivateKey: ReadableFromBuffer {
+    static func read(from buffer: inout ByteBuffer) throws -> Self {
+        guard
+            let publicKeyLength = buffer.readInteger(as: UInt32.self),
+            let publicKey = buffer.readBytes(length: Int(publicKeyLength))
+        else {
+            throw InvalidKey()
+        }
+
+        guard
+            let privateKeyLength = buffer.readInteger(as: UInt32.self),
+            let privateKey = buffer.readBytes(length: Int(privateKeyLength)),
+            Array(privateKey[32...]) == publicKey
+        else {
+            throw InvalidKey()
+        }
+        
+        return try Self.init(rawRepresentation: privateKey[..<32])
+    }
+}
+
+enum OpenSSH {
+    enum Cipher: String {
+        case none
```

Comment:
Can we add a `decrypt` function here?

`func decryptBuffer(_ buffer: inout ByteBuffer) throws`

We'll also need some parameters in addition to the buffer for the decryption key and IV.

Even though `Cipher.none` wouldn't do any actual decryption, adding the infrastructure for decrypting is a great start for when we do get to it.

Reply:
Btw; this is a non-blocker to me, since the API is all internal.

Reply:
Just to check what cause I'm not sure for the openSSH format do you know which parts of the format are then encrypted? (ie what buffer to call decrypt on) are the length encodings included in the encrypted data?